File: modules/pitch-input/android/src/main/java/expo/modules/pitchinput/PitchInputModule.kt
package expo.modules.pitchinput

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Handler
import android.os.HandlerThread
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.nio.ByteBuffer
import java.nio.ByteOrder

private const val RATE = 44100

/** Frames per analysis window. Must match WINDOW in src/lib/tuner-math.ts. */
private const val WINDOW = 4096

class MicUnavailableException(cause: String) :
  CodedException("ERR_MIC_UNAVAILABLE", "Microphone unavailable: $cause", null)

/**
 * A microphone tap and nothing else.
 *
 * Pitch detection deliberately lives in src/lib/tuner-math.ts instead of here,
 * so the algorithm is written once, in one language, and can be tested on a
 * desktop without a device. This class owns a ring buffer of the newest WINDOW
 * frames and hands them to JS whenever it asks.
 */
class PitchInputModule : Module() {
  private var recorder: AudioRecord? = null
  private var thread: HandlerThread? = null

  @Volatile
  private var running = false

  // The ring. Guarded by `lock` because the capture thread writes it while the
  // JS thread reads it.
  private val ring = ShortArray(WINDOW)
  private var writeIndex = 0
  private var filled = false
  private val lock = Any()

  override fun definition() = ModuleDefinition {
    Name("PitchInput")

    Property("sampleRate") { RATE }
    Property("isRecording") { running }

    Function("start") { start() }
    Function("stop") { stop() }

    Function("read") {
      synchronized(lock) {
        if (!filled) {
          ByteArray(0)
        } else {
          // Little-endian Int16, oldest first, so JS can view it as an Int16Array.
          val out = ByteBuffer.allocate(WINDOW * 2).order(ByteOrder.LITTLE_ENDIAN)
          for (i in 0 until WINDOW) out.putShort(ring[(writeIndex + i) % WINDOW])
          out.array()
        }
      }
    }

    OnDestroy { stop() }
  }

  private fun start() {
    if (running) return

    val minBuf = AudioRecord.getMinBufferSize(
      RATE,
      AudioFormat.CHANNEL_IN_MONO,
      AudioFormat.ENCODING_PCM_16BIT
    )
    if (minBuf <= 0) throw MicUnavailableException("no buffer size")

    // VOICE_RECOGNITION, not VOICE_COMMUNICATION: the latter applies AGC and
    // noise suppression, which colour the very pitch we are trying to measure.
    val rec = try {
      AudioRecord(
        MediaRecorder.AudioSource.VOICE_RECOGNITION,
        RATE,
        AudioFormat.CHANNEL_IN_MONO,
        AudioFormat.ENCODING_PCM_16BIT,
        maxOf(minBuf, WINDOW * 4)
      )
    } catch (e: Exception) {
      throw MicUnavailableException(e.message ?: "constructor failed")
    }

    if (rec.state != AudioRecord.STATE_INITIALIZED) {
      rec.release()
      throw MicUnavailableException("not initialized")
    }

    synchronized(lock) {
      writeIndex = 0
      filled = false
    }

    recorder = rec
    running = true
    rec.startRecording()

    val t = HandlerThread("pitch-input").also { it.start() }
    thread = t
    Handler(t.looper).post {
      val chunk = ShortArray(1024)
      while (running) {
        val n = rec.read(chunk, 0, chunk.size)
        if (n <= 0) continue
        synchronized(lock) {
          for (i in 0 until n) {
            ring[writeIndex] = chunk[i]
            writeIndex = (writeIndex + 1) % WINDOW
            if (writeIndex == 0) filled = true
          }
        }
      }
    }
  }

  private fun stop() {
    if (!running) return
    running = false
    thread?.quitSafely()
    thread = null
    recorder?.let {
      runCatching { it.stop() }
      it.release()
    }
    recorder = null
  }
}

File: modules/pitch-input/ios/PitchInputModule.swift
import AVFoundation
import ExpoModulesCore

/** Frames per analysis window. Must match WINDOW in src/lib/tuner-math.ts. */
private let WINDOW = 4096

/**
 * Microphone tap for the tuner. Mirrors the Android module exactly: a ring of
 * the newest WINDOW frames, handed to JS as little-endian Int16 PCM, oldest
 * first. All pitch detection lives in src/lib/tuner-math.ts.
 *
 * NOTE: written without a device to verify against — this project has no Apple
 * developer account yet. Test before trusting it.
 */
public class PitchInputModule: Module {
  private let engine = AVAudioEngine()
  private var running = false
  // The hardware rate, not an assumption: iOS routinely hands back 48 kHz.
  private var rate: Double = 44100
  private var ring = [Int16](repeating: 0, count: WINDOW)
  private var writeIndex = 0
  private var filled = false
  private let lock = NSLock()

  public func definition() -> ModuleDefinition {
    Name("PitchInput")

    Property("sampleRate") { self.rate }
    Property("isRecording") { self.running }

    Function("start") { try self.start() }
    Function("stop") { self.stop() }

    Function("read") { () -> Data in
      self.lock.lock()
      defer { self.lock.unlock() }
      guard self.filled else { return Data() }
      var out = Data(capacity: WINDOW * 2)
      for i in 0..<WINDOW {
        var s = self.ring[(self.writeIndex + i) % WINDOW].littleEndian
        withUnsafeBytes(of: &s) { out.append(contentsOf: $0) }
      }
      return out
    }

    OnDestroy { self.stop() }
  }

  private func start() throws {
    guard !running else { return }

    let session = AVAudioSession.sharedInstance()
    // .measurement leaves the signal alone — no AGC or processing shaping the
    // pitch we are trying to measure.
    try session.setCategory(.playAndRecord, mode: .measurement, options: [.defaultToSpeaker])
    try session.setActive(true)

    lock.lock()
    writeIndex = 0
    filled = false
    lock.unlock()

    let input = engine.inputNode
    let format = input.outputFormat(forBus: 0)
    rate = format.sampleRate

    input.installTap(onBus: 0, bufferSize: 1024, format: format) { [weak self] buffer, _ in
      guard let self, let channel = buffer.floatChannelData?[0] else { return }
      self.lock.lock()
      for i in 0..<Int(buffer.frameLength) {
        let clamped = max(-1, min(1, channel[i]))
        self.ring[self.writeIndex] = Int16(clamped * 32767)
        self.writeIndex = (self.writeIndex + 1) % WINDOW
        if self.writeIndex == 0 { self.filled = true }
      }
      self.lock.unlock()
    }

    engine.prepare()
    try engine.start()
    running = true
  }

  private func stop() {
    guard running else { return }
    running = false
    engine.inputNode.removeTap(onBus: 0)
    engine.stop()
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
  }
}
